### DISCUSSION TITLE
Tailwind CSS

### DISCUSSION THREAD

**Discussion by ghost** [tamagui/tamagui#114] [Q&A]
Does anyone know if it's possible to adopt both tamagui and tailwind? Wondering if there's a way to not re-create existing tailwind styles for the transition.

**Comment by natew** [comment]
If you're transitioning, keep tailwind around while you port things over. You can try matching some shorthands, but I don't recommend trying to "wrap" tamagui to adapt it closer as you'll lose some compilation.

**Reply by redbar0n** [reply]
Would a default set of Tailwind aligned shorthands for Tamagui make sense?

For faster onboarding of people coming from Tailwind (and have already got the Tailwind API in their fingers).

**Reply by natew** [reply]
That would be a great PR to send, happy to accept. Can just duplicate the shorthand’s and change them.

**Comment by redbar0n** [comment]
I'm not using Tailwind myself, but curious about this issue, since I think smooth sailing for those who do would increase Tamagui's adoption (which would be good for the ecosystem).

I realized that the experience with [shorthands](https://tamagui.dev/docs/intro/introduction#shorthands) wouldn't be the same, since the Tamagui shorthands API is different from Tailwind.

A [brief demonstration by Theo](https://www.youtube.com/watch?v=rUqLVNr5oFU) of why people seem to love the _typing experience_ of Tailwind, in summary: it avoids the intermediary `="  "` signs that are harder to reach on the keyboard.

[create-t3-app](https://github.com/t3-oss/create-t3-app#info) is [getting quite popular](https://star-history.com/#t3-oss/create-t3-app&Date) on web (5.3k stars). Near the end of August [they made](https://www.jumr.dev/blog/t3-turbo) a turborepo template called [create-t3-turbo](https://github.com/t3-oss/create-t3-turbo) as a template monorepo with Expo to _"consume the tRPC API on both web and mobile"_. It is already [trending upwards](https://star-history.com/#t3-oss/create-t3-turbo&Date) as well. The only difference from the excellent [zART-stack](https://github.com/trpc/zart) (that for some reason [didn't catch on that much](https://star-history.com/#trpc/zart&t3-oss/create-t3-turbo&Timeline)), besides Turborepo, is that create-t3-turbo uses Tailwind via [NativeWind](https://github.com/marklawlor/nativewind). Though popularity may have many coincidences, ofc.

But I'd like to point out that people [seem to be **really** keen](https://moiva.io/?npm=nativewind+react-native-tailwind+react-native-tailwind-classnames+react-native-tailwindcss+react-native-zephyr+tailwind-react-native-classnames+tailwind-rn+tailwindcss-react-native) on getting Tailwind to Native ([crazy many libraries for the same purpose](https://github.com/jaredh159/tailwind-react-native-classnames/discussions/171#discussioncomment-3714956)). You've also seen the interest coming up from time to time in the Discord chat. And as far as component libraries on web go, ChakraUI [receives some love](https://github.com/chakra-ui/chakra-ui/issues/634#issuecomment-619626792) for being compatible with Tailwind. 

Tailwind now has 60k stars, with several attempts to port it to native. So it seems like it's here to stay for a while. I think Tamagui could get a decent boost from it. Heck, Tamagui could maybe even get a boost for pure webapps, since [Twind](https://twind.dev/handbook/introduction.html#features) has become quite popular (2.7k stars), and it does at runtime what Tamagui seems to be doing already at compile time. What people dislike with Tailwind is the _"need for PostCSS, configuration, purging, or autoprefixing"_. If Tamagui already takes care of all that, without extra config, AND has crossplatform support, then it's an immediate win.

With the shorthands and the compiler, it seems so close to achieving it. But as far as I can tell, there is the question of:

### How to make it compatible with typed inline styles, if one puts all the utility styles inside a single `=" "` ?

Tamagui can't use `className` since the docs say it's [web only](https://tamagui.dev/docs/intro/props#tamagui-props).

Some potential solutions:

1. Tamagui compiles Tailwind /  NativeWind styling. Issue is now Tamagui has 2 different API's. Could they be mixed and matched? What about maintenance? What about annoyances when jumping from one Tailwind project to another?

2. VS Code extension that expands Tailwind tags to the Tamagui API. The Tailwind aficionados could use their automatized Tailwind motor skills to _type Tailwind but see Tamagui's style props_. Good for onboarding to Tamagui's API. But users would only gain the typing experience of Tailwind in Tamagui if they learn the Tailwind API  /  classnames.

3. Tamagui offers its own _Tailwind-similar_ typing experience, where you could get into a _styling flow_, not having to type all the `=" "`...  But how could it be typed properly? Maybe a VS Code extension could solve this for the existing Tamagui API? That way, people could style as fast as Tailwind, but within the confines of the existing Tamagui API. (Doesn't help the people who know the Tailwind API and want to reuse that knowledge, though. But they may use option 2, perhaps.)

**Reply by natew** [reply]
@redbar0n this made me realize we *can* do something...

https://github.com/tamagui/tamagui/pull/234

If someone wants to fill in the rest of these values I'll happily merge it and we can add some docs.

Would allow for:

```tsx
import { Stack } from '@tamagui/tailwind'

export default () => <Stack w-0 />
```

**Reply by redbar0n** [reply]
@Ashley @wezter96 @hirbod @dohomi @Zetax  @AngelRs

**Reply by redbar0n** [reply]
For those interested, this is how an example with Tamagui Tailwind would look with the proposed API.

```tsx
// Tamagui Tailwind, set to correspond with TailwindCSS API: https://tailwindcss.com/docs/
function Label({ text }: { text: string }) {
  return (
    <XStack h-6 px-2 py-1 bg-$background mr-2 rounded-r items-center>
      <YStack mr-1>
        <Icon />
      </YStack>
      <Text text-xs font-bold leading-4 text-ellipsis>
        {text}
      </Text>
    </XStack>
  )
}
// leading-4 used above because:
// lineHeight={14} in Tamagui. No close match found in Tailwind.
// leading-3 in Tailwind is: line-height: .75rem; /* 12px */
// leading-4 in Tailwind is: line-height: 1rem; /* 16px */

// Tamagui, for comparison. From: https://github.com/tamagui/tamagui/blob/bcd682447a7009e1bd6d9805d6e50e7587dba995/starters/benchmark/packages/app/features/user/BenchTamagui.tsx#L56
// With current actual Tamagui shorthands from: https://github.com/tamagui/tamagui/blob/master/packages/shorthands/src/index.tsx
function Label({ text }: { text: string }) {
  return (
    <XStack h={6} px={2} py={1} bc="$background" mr={2} borderRadius={4} alignItems="center">
      <YStack mr={1}>
        <Icon />
      </YStack>
      <Text fontSize="xs" fontWeight="bold" lineHeight={14} ellipse>
        {text}
      </Text>
    </XStack>
  )
}

// NativeBase, for comparison. From:  https://github.com/tamagui/tamagui/blob/bcd682447a7009e1bd6d9805d6e50e7587dba995/starters/benchmark/packages/app/features/user/BenchNativeBase.tsx#L128
function Label({ text }: { text: string }) {
  return (
    <HStack h={6} px={2} py={1} bg="brand.black" mr={2} borderRadius="4" alignItems="center">
      <Box mr="1">
        <Icon />
      </Box>
      <Text fontSize="xs" fontWeight="bold" lineHeight="14px" isTruncated>
        {text}
      </Text>
    </HStack>
  )
}

// React Native inline style objects, for comparison. From: https://github.com/tamagui/tamagui/blob/bcd682447a7009e1bd6d9805d6e50e7587dba995/starters/benchmark/packages/app/features/user/BenchReactNative.tsx#L109
function Label({ text }: { text: string }) {
  return (
    <View
      style={{
        height: 24,
        paddingHorizontal: 8,
        paddingVertical: 4,
        backgroundColor: '#000',
        marginRight: 8,
        borderRadius: 4,
        alignItems: 'center',
        flexDirection: 'row',
      }}
    >
      <View
        style={{
          marginRight: 4,
        }}
      >
        <Icon />
      </View>
      <Text
        style={{
          fontSize: 12,
          fontWeight: '700',
          lineHeight: 14,
          color: 'white',
        }}
      >
        {text}
      </Text>
    </View>
  )
}
```

Update: here is [a more extended Tamagui code example](https://github.com/tamagui/tamagui/blob/master/apps/site-vite/src/components/home/HeroResponsive.client.tsx). One should likely try to translate such an example to the proposed Tamagui Tailwind API, to see if everything is translateable.

**Reply by dohomi** [reply]
Would it be maybe better to keep the TailwindCSS props behind a `className` similar on how the syntax works with Nativewind?

**Reply by redbar0n** [reply]
> Would it be maybe better to keep the TailwindCSS props behind a `className` similar on how the syntax works with Nativewind?

I’m wondering about the same..

Because otherwise, won’t there be potential name clashes with other props (belonging to Tamagui or the user), @natew ? 

`className` is forbidden with RNW, if I [recall](https://github.com/tamagui/tamagui/commit/34768a9eb7abdd3c9870b41446315c2b46796795#commitcomment-64756874) correctly. So maybe a `tw` prop?

**Reply by natew** [reply]
Putting it on className you lose types and lose easily passing/merging props down. It could be part of “tw” as a helper for directly transitioning easier (variant function), but it’s a big downside in usability, safety, features, and performance.

**Reply by natew** [reply]
Your Tamagui example isn’t using some shorthand’s it could and technically would want to use tokens for m=“$2” which is easier to type

**Reply by natew** [reply]
I think the types on values and flexibility on values is with the slight extra syntax, but happy to support this as an option

**Reply by dohomi** [reply]
Alternatively, if Tamagui would respect Nativewind components then there wouldn't be any further action needed on the development of Tamagui. Would that be an option?

**Reply by natew** [reply]
What do you mean by respect?

**Reply by dohomi** [reply]
Sorry wrong wording: I meant if users can include Nativewind and style their components with both, Nativewind and Tamagui then there isn't much overhead.
Example how this would look like with StyledComponennt
```ts
import { StyledComponent } from "nativewind";
import {styled, Text} from "tamagui"

const CustomText = styled(Text,{
// some component styles...
})

export function MyComponent() {
  return (
    <>
      <StyledComponent component={CustomText} tw="font-bold">
        Hello world
      </StyledComponent>
      <StyledComponent component={CustomText} className="font-bold">
        Hello world
      </StyledComponent>
    </>
  );
}
```
Or using styled from Nativewind:
```ts
import { StyledComponent, styled: nativeStyled } from "nativewind";
import {styled, Text} from "tamagui"

const NativeText = nativeStyled(Text, 'font-bold')

const CustomText = styled(NativeText,{
// some component styles...
})


```

**Reply by redbar0n** [reply]
@natew 

> I think the types on values and flexibility on values **is with** the slight extra syntax, but happy to support this as an option

is what?

> It could be part of “tw” as a helper for directly transitioning easier (variant function), but it’s a big downside in usability, safety, features, and performance.

Could you explain briefly why a `tw` prop would have those downsides?

`tw` prop:

```tsx
<XStack tw={"h-6 px-2 py-1 bg-$background mr-2 rounded-r items-center"}>
```

Is it because of difficulties with type inference in string interpolation?

Alternatively, with Template Literal,  which [twind recommends](https://twind.dev/handbook/styling-with-twind.html#the-tw-function):

```tsx
<XStack tw`h-6 px-2 py-1 bg-$background mr-2 rounded-r items-center`>
```

Or maybe both, for familiarity:

```tsx
<XStack tw={`h-6 px-2 py-1 bg-$background mr-2 rounded-r items-center`}>
```

**Reply by natew** [reply]
*is better with

- Prettier wont format that so you're stuck with giant horizontal lines
- Lose out on value types, and stuff like "you duplicated this property"
- Have to write extra compared to flat
- Performance just due to tamagui not supporting this first class you'd end up with very complex variant definition that may not flatten or compile well

A few other things. 

I think it's better to avoid style props on high level components if you want their API to be custom. That pattern should be extremely common, so flat style props don't in practice ever collude the namespace.

**Reply by natew** [reply]
I think another (really) big thing is being able to use objects:

- Pass in styles to child, Child can then de-structure, change values, etc very easily
- Object spreading likewise is nice and gives TS warnings if you're overwriting, etc
- Separate hover/press/etc means you can control those easily with destructure/rest/spread
- Separate media queries means you can control those easily with destructure/rest/spread

**Reply by nandorojo** [reply]
I agree with all those points @natew. This has been my draw to object-style design libraries.

I'm considering playing with Nativewind / Tamagui for a new project we're working on at BeatGig. While I have my concerns about the DX and typesafety of Tailwind, I like the fact that it has the compiler performance benefits with a seemingly lighter configuration than Tamagui. That said, I'd like to try out both in conjunction somehow and see how it goes if possible.

**Reply by redbar0n** [reply]
Some of [Theo's arguments against Tailwind](https://youtu.be/-R9uHm0G5m4?t=406):

1. Need to parse all files that MIGHT use Tailwind classes (can get rough for large codebases, but not too rough, usually fast)
2. Need to know or learn CSS [Ed.: might not be what React Native developers want]
3. Codesplitting isn't really a thing :(

I think Tamagui could improve on Tailwind here.

There is tremendous momentum ([61.1k stars currently](https://star-history.com/#tailwindlabs/tailwindcss&tamagui/tamagui&marklawlor/nativewind&Date), [stateofcss2021  rankings](https://2021.stateofcss.com/en-US/technologies/css-frameworks/#css_frameworks_experience_ranking), [Tailwind approaching Bootstrap in NPM downloads](https://moiva.io/?npm=bootstrap+tailwindcss)) behind Tailwind in the ecosystem right now, and a lot of developers will simply choose it because of that momentum. They won't necessarily look into, or educate themselves, on these finer points you've raised. The simplest thought process being: "Does it support Tailwind? Check". If that barrier is out of the way then the discussion can move to finer points, than trying to argue to the potential adopter why Tamagui's way is superior to Tailwind's. So I think it could be worth it for Tamagui to tap into some of Tailwind's momentum and leverage that for its own. Then, people could be slowly guided towards an even better way (like style objects, if that is it). Because educating the whole market would take time and effort, and could be an uphill battle. The market wants what the market wants.

Why not leverage that existing awareness and interest?

It's not necessarily the only thing that matters, ofc, but imho, it's a point worth taking into account.

PS: Some have argued that [the key to Tailwind](https://mxstbr.com/thoughts/tailwind)'s success (where other atomic css approaches have failed) is that it _lowers skill barrier_ due to its carefully curated set of constraints, which give good guardrails. (This is opposed to free-form styling with style objects that seniors may prefer.)

**Reply by nandorojo** [reply]
I think the momentum is a good argument to use Tailwind. Besides the fact that the APIs are nice, I’m drawn to its maintenance stability. 

It’s a seemingly-profitable company with a long track record and many users. I wouldn’t want to underrate that as a reason people use it, especially since I’ve experienced what it’s like to invest in someone else’s project, only for it to go stale.

In any case, I agree that it would be smart to ride the wave a bit.

**Reply by jonsherrard** [reply]
@redbar0n out of genuine interest: what do you foresee getting from Tamagui using TW syntax that you wouldn't get from using Tailwind itself and Nativewind on RN https://www.nativewind.dev/quick-starts/solito?

**Reply by redbar0n** [reply]
> @redbar0n out of genuine interest: what do you foresee getting from Tamagui using TW syntax that you wouldn't get from using Tailwind itself and Nativewind on RN https://www.nativewind.dev/quick-starts/solito?

The benefits Tamagui brings over NativeWind. Mainly the included component library, support for animations, and the optimizing compiler. @natew can probably fill in with more benefits.

**Reply by redbar0n** [reply]
@nandorojo what did you end up choosing between Nativewind and Tamagui, and what has been your experience?

**Reply by redbar0n** [reply]
FFR: valueless attributes in Tailwind successors UnoCSS and WindiCSS’s Attributify Mode https://antfu.me/posts/reimagine-atomic-css#attributify-mode
is similar to @natew’s suggested API.

**Reply by corysimmons** [reply]
I didn't read all this thread (getting bored of the "why tailwind?" debate), but you'd certainly want to wrap the Tailwind-specific stuff in its own prop for readability alone:

`<View p-2 flex-1 onPress={thing} bg-red-500 isOpen />` vs

`<View tw="p-2 flex-1 bg-red-500" onPress={thing} isOpen />`

You can write a VSCode plugin to hint at values within a prop string. You can even split/sort those values. [Here](https://github.com/tailwindlabs/prettier-plugin-tailwindcss/blob/f9b249b171003da870e6865fd2b54c226b501a97/src/index.js#L477) is the line where the prop is defined in their official Prettier plugin. You could probably fork/patch that with the prop you want (e.g. `tw`) and it'd work.

There are a couple of reasons almost everyone likes Tailwind, and styled-system and other patterns have pretty much died, but the primary one is Tailwind's team has spent years making a real & massive design system:
- nice default classes
  - colors
  - spacing
  - fonts
  - etc
- intuitive API
- good docs
- editor plugins to do things like hint and sort classnames
- thriving/active/self-funded maintenance
- actually looks good vs every other default system I've ever seen

**Reply by natew** [reply]
I think Tamagui beats Tailwind in every category listed though, better types, better autocomplete, way less weird stuff, no very arcane DSL over CSS, merges downwards properly, themes, etc. Comes with sane defaults but also more flexible to customize. 

Flat props are a matter of preference too, I find your first example far more nice, but that’s just me. I don’t recommend strings for styles, you lose types, destructuring, re-ordering, merge sanity going downwards. 

Actually Tamagui considered Tailwind before it was created and decided against it for many reasons. But yes for familiarity we can have them as an option.

**Comment by redbar0n** [comment]
Both Nativewind and Tamagui has had a lot of growth recently: https://moiva.io/?npm=@tamagui/core+nativewind

Interestingly, Nativewind has twice the amount of downloads, but Tamagui twice the amount of Github stars. Not sure what that indicates, and which generally follows which. Some download, then star, presumably, but others star more than they download.

**Reply by corysimmons** [reply]
People don't realize they need both so they reach for NativeWind to design something, then they can't Popover and start looking for stuff.

It'd be great if Tamagui changed to almost exclusively focus on being the HeadlessUI of the ecosystem. 😁 

The Stars are probably because of Fernando pushing people to Tamagui in his latest conference presentation, but all this is just guessing. 🤷

**Reply by natew** [reply]
Need both? You don’t need NativeWind at all with Tamagui and I’d recommend against mixing them as they overlap completely. 

Tamagui components build on top of Tamagui core and there’s really no way around that.

**Reply by nandorojo** [reply]
> The Stars are probably because of Fernando pushing people to Tamagui in his latest conference presentation, but all this is just guessing. 🤷

For what it’s worth, I talked about both Nativewind and Tamagui in that talk. Nativewind also got a full release and section in the v2 docs of Solito. So it’s definitely gotten its shine. But Tamagui is taking on a much wider scope.

I like the idea of Tamagui being Headless UI (and that’s becoming an option), but you do miss out on so many necessary wins it provides for building a scalable UI that stays fast on Web.

**Comment by redbar0n** [comment]
FWIW, the comments on Theo’s recent StyleX video are a good insight/highlight of how much people prefer the Tailwind writing ergonomics (over style objects): https://youtu.be/v4HkkjXmEMM?si=uMiSG_9Cwud8u5WE

**Reply by nandorojo** [reply]
Yeah but StyleX is way slower than tamagui to use. It’s like stylesheet. You have to name everything and write it outside of render. I don’t think objects themselves (or inline props) vs strings is the issue. The killer feature is writing styles inline and being able to copy + paste them easily.

**Comment by redbar0n** [comment]
StyleX example of tailwind parser at the end: 
https://t.co/mz9Mzw17rW

Define typesafe tw classname strings in JS.

Maybe tamagui could leverage that open source code there, for short term gains for those coming from tailwind and just want their tw tags into JS / Tamagui.
